Added Custom callbacks

phyzical opened the pull request:
Hey @RoccoC !

Love the package, but i had a inkling to use custom sounds and rather than going down the routes of making another plugin for webpack just "just fire a snippet" after your code does its thing i decided to just add the ability.

Feel free to decline if this isnt something you want in your package, or you have  concerns with what ive submitted let me know and i can improve the implementation.

If you dont want this functionally no worries, i can just continue using my fork.

summary:
* add functionality to provide a base callback, and a callback per type of notification with the ability to provide overriding false if using a base callback (tried to mimic the sounds settings).
* updated tests
* updated readme

and an example of new usage playing a custom sound on complete
```
const player = require('play-sound')(opts = {});
let plugins = [];

if (processEnv.NODE_ENV === 'development') {
    plugins = plugins.concat([
        new WebpackBuildNotifierPlugin({
            title: " Build Notification",
            logo: path.join(__dirname,'src/icons/favicon.png'),
            sound: false,
            suppressCompileStart: false,
            customCallback: () => {
                player.play(path.join(__dirname,'allfinished.mp3'), (err) => {
                    if (err) console.log(err)
                });
            },
            compilationCustomCallback: false
        }),
    ]);
```

Thanks!

RoccoC commented:
Hi @phyzical , thanks for the PR! Definitely open to adding this functionality. Give me a day or two to review. :) Thanks again!

RoccoC commented on index.d.ts:
```diff
@@ -50,6 +50,30 @@ declare namespace WebpackBuildNotifierPlugin {
+     * The default function to be run after the compilation notify has fired
+     * Set to false to fire no function for compilation notifications. Takes precedence over the *customCallback* configuration option.
+     */
+    compilationCustomCallback?: Function | false;
```
Rather than adding a callback for each scenario, what would you think about consolidating to two callbacks: one for `onCompileStart`, and another which accepts a parameter to indicate the status of the completed build? Maybe something like this?

```
type CompilationStatus = 'success' | 'error' | 'warning';

onCompileStart?: (compilation: webpack.compilation.Compilation) => void;
onComplete?: (status: CompilationStatus, compilation: webpack.compilation.Compilation) => void;
```

I feel that the config is a bit bloated as it is, so this would help keep the surface area a bit more tidy. :)

phyzical replied:
```updated the pr,
i'm not sure how type hinting goes down in JS though but i did it for TS)

removed the ability to pass false and it just calls the default OR the function provided and moved the functionality of CompilationStatus checking onto the implementation provided to the completeCallback.

Hopefully this is what you were after?

Thanks!```

phyzical commented:
updated the pr, 
i'm not sure how type hinting goes down in JS though but i did it for TS)

removed the ability to pass false and it just calls the default OR the function provided and moved the functionality of CompilationStatus checking onto the implementation provided to the completeCallback.

Hopefully this is what you were after?

Thanks!

phyzical commented:
@RoccoC  bump :)

RoccoC commented:
Looks great! I merged into `release/callbacks` and will reintegrate with master and publish a release shortly. Thanks again for the contribution!

RoccoC commented:
Released `1.1.0`. Thanks again!

phyzical commented:
Thanks dude!

confirmed 1.1.0 as working 👍 

farewell the next time our paths cross 😄
